File: ruby/roman-numerals/roman.rb
class Fixnum
  
  def to_roman
    i = self
    s = ''
    ROMAN_MAPPING.each do |roman, arabic|
      while i >= arabic
        s << roman.to_s.upcase
        i -= arabic
      end
    end
    s
  end

  private

  ROMAN_MAPPING = {
    m:  1000,
    cm: 900,
    d:  500,
    cd: 400,
    c:  100,
    xc: 90,
    l:  50,
    xl: 40,
    x:  10,
    ix: 9,
    v:  5,
    iv: 4,
    i:  1
  }
end

File: elixir/roman-numerals/roman.exs
defmodule Roman do
  @mapping [
    { 1000, "M" },
    { 900,  "CM" },
    { 500,  "D" },
    { 400,  "CD" },
    { 100,  "C" },
    { 90,   "XC" },
    { 50,   "L" },
    { 40,   "XL" },
    { 10,   "X" },
    { 9,    "IX" },
    { 5,    "V" },
    { 4,    "IV" },
    { 1,    "I" }
  ]
  @doc """
  Convert the number to a roman number.
  """
  @spec numerals(pos_integer) :: String.t
  # Using the bodyless function declaration the accumulator is set to an empty
  # binary string. Then, on compile time, functions are generated for each step
  # in the roman mapping. It start at the highest number and checks if it can
  # be subtracted from the given number without falling below zero. If it
  # matches the corresponding roman letter is added to the accumulator and it
  # recursively continues untill reaching zero.
  def numerals(n, acc \\ "")
  def numerals(0, acc), do: acc
  for { arabic, roman } <- @mapping do
    def numerals(n, acc) when n - unquote(arabic) >= 0 do
      numerals(n - unquote(arabic), acc <> unquote(roman))
    end
  end
end